core/shared/src/main/scala/scalaz/zio/stream/Stream.scala

```diff
@@ -0,0 +1,733 @@
+      fa.flatMap(f(s, _))
+  }
+
+  /**
+   * Flattens a stream of streams into a stream, by concatenating all the
+   * substreams.
+   */
+  final def flatten[E, A](fa: Stream[E, Stream[E, A]]): Stream[E, A] =
+    fa.flatMap(identity)
+
+  /**
+   * Unwraps a stream wrapped inside of an `IO` value.
+   */
+  final def unwrap[E, A](stream: IO[E, Stream[E, A]]): Stream[E, A] =
+    new Stream[E, A] {
+      override def fold[E1 >: E, A1 >: A, S](s: S)(f: (S, A1) => IO[E1, Step[S]]): IO[E1, Step[S]] =
+        stream.flatMap(_.fold[E1, A1, S](s)(f))
+    }
+
+  /**
+   * Constructs a stream from a resource that must be acquired and released.
+   */
+  final def bracket[E, A, B](
+    acquire: IO[E, A]
+  )(release: A => IO[Nothing, Unit])(read: A => IO[E, Option[B]]): Stream[E, B] =
+    managed(Managed(acquire)(release))(read)
+
+  final def managed[E, A, B](m: Managed[E, A])(read: A => IO[E, Option[B]]) =
+    new Stream[E, B] {
+      override def fold[E1 >: E, B1 >: B, S](s: S)(f: (S, B1) => IO[E1, Step[S]]): IO[E1, Step[S]] =
+        m use { a =>
+          read(a).flatMap {
+            case None    => IO.now(Step.Stop(s))
+            case Some(b) => f(s, b)
+          }
+        }
+    }
+
+  /**
+   * Constructs an infinite stream from a `Queue`.
+   */
+  final def fromQueue[A](queue: Queue[A]): Stream[Nothing, A] =
+    unfoldM(())(_ => queue.take.map(a => Some((a, ()))))
+
+  /**
+   * Constructs a stream from effectful state. This method should not be used
+   * for resources that require safe release. See `Stream.fromResource`.
+   */
+  final def unfoldM[S, E, A](s: S)(f0: S => IO[E, Option[(A, S)]]): Stream[E, A] =
+    new Stream[E, A] {
+      override def fold[E1 >: E, A1 >: A, S2](s2: S2)(f: (S2, A1) => IO[E1, Step[S2]]): IO[E1, Step[S2]] = {
+        def loop(s: S, s2: S2): IO[E1, Step[(S, S2)]] =
+          f0(s).flatMap {
+            case None         => IO.now(Step.Stop((s, s2)))
+            case Some((a, s)) => f(s2, a).map(_.map(s2 => (s, s2)))
+          }
+
+        loop(s, s2).map(_.map(_._2))
+      }
+    }
+
+  /**
+   * Constructs a stream from state.
+   */
+  final def unfold[S, A](s: S)(f0: S => Option[(A, S)]): Stream[Nothing, A] =
+    new StreamPure[A] {
+      override def fold[E, A1 >: A, S2](s2: S2)(f: (S2, A1) => IO[E, Step[S2]]): IO[E, Step[S2]] = {
+        def loop(s: S, s2: S2): IO[E, Step[(S, S2)]] =
+          f0(s) match {
+            case None         => IO.now(Step.Stop((s, s2)))
+            case Some((a, s)) => f(s2, a).map(_.map(s2 => (s, s2)))
+          }
+
+        loop(s, s2).map(_.map(_._2))
+      }
+
+      override def foldPure[A1 >: A, S2](s2: S2)(f: (S2, A1) => Step[S2]): Step[S2] = {
+        def loop(s: S, s2: S2): Step[(S, S2)] =
+          f0(s) match {
+            case None         => Step.Stop((s, s2))
+            case Some((a, s)) => f(s2, a).map(s2 => (s, s2))
+          }
+
+        loop(s, s2).map(_._2)
+      }
+    }
```

Comment:
It seems that the `unfold` stop after the first value. The `loop` is not called recursively.

I tried the following code which seems to work better:

    final def unfold[S, A](s: S)(f0: S => Option[(A, S)]): Stream[Nothing, A] =
      new StreamPure[A] {
        override def fold[E, A1 >: A, S2](s2: S2)(f: (S2, A1) => IO[E, Step[S2]]): IO[E, Step[S2]] = {
          def loop(s: S, s2: S2): IO[E, Step[S2]] =
            f0(s) match {
              case None         => IO.now(Step.Stop(s2))
              case Some((a, s)) => f(s2, a).flatMap {
                case Step.Cont(s2) => loop(s, s2)
                case Step.Stop(s2) => IO.now(Step.stop(s2))
              }
            }
  
          loop(s, s2)
        }
  
        override def foldPure[A1 >: A, S2](s2: S2)(f: (S2, A1) => Step[S2]): Step[S2] = {
          def loop(s: S, s2: S2): Step[S2] =
            f0(s) match {
              case None         => Step.Stop(s2)
              case Some((a, s)) => f(s2, a) match {
                case Step.Cont(s2) => loop(s, s2)
                case Step.Stop(s2) => Step.stop(s2)
              }
            }
  
          loop(s, s2)
        }
      }

The problem is the same with `unfoldM`.

Reply:
🤦🏼‍♂️